samples/matter/common/src/bridge/ble_connectivity_manager.cpp

```diff
@@ -444,30 +448,103 @@ CHIP_ERROR BLEConnectivityManager::Scan(ScanDoneCallback callback, void *context
 	return ret;
 }
 
+void BLEConnectivityManager::Recovery::RemoveRecovered(BLEBridgedDeviceProvider *provider)
+{
+	if (!provider) {
+		return;
+	}
+
+	sys_snode_t *node = sys_slist_peek_head(&Instance().mRecovery.mListToRecover);
+	ListItem *item;
+
+	/* Iterate through the list of providers and remove the requested one. */
+	do {
+		item = CONTAINER_OF(node, __typeof__(*item), mNode);
+
+		if (!item) {
+			return;
+		}
+
+		if (item->mProvider == provider) {
+			sys_slist_find_and_remove(&Instance().mRecovery.mListToRecover, node);
+			Platform::Delete(item);
+			return;
+		}
+		node = sys_slist_peek_next(node);
+	} while (node != nullptr);
+}
+
+bool BLEConnectivityManager::Recovery::IsProviderToRecover(BLEBridgedDeviceProvider *provider)
+{
+	if (!provider || !IsNeeded()) {
+		return false;
+	}
+
+	sys_snode_t *node = sys_slist_peek_head(&Instance().mRecovery.mListToRecover);
+	ListItem *item;
+
+	/* Iterate through the list of providers and check if the given one is on it. */
+	do {
+		item = CONTAINER_OF(node, __typeof__(*item), mNode);
+
+		if (!item) {
+			return false;
+		}
+
+		if (item->mProvider == provider) {
+			LOG_ERR("Provider found on a list to recover");
+			return true;
+		}
+		node = sys_slist_peek_next(node);
+	} while (node != nullptr);
+
+	return false;
+}
+
 void BLEConnectivityManager::ReScanCallback(ScannedDevice *devices, uint8_t count, void *context)
 {
-	/* Get a provider form the recovery ring buffer, if there is none entries it should be nullptr. */
-	BLEBridgedDeviceProvider *provider = Instance().mRecovery.GetProvider();
-	/* check if device still need to be found (for example it could be removed from the bridge devices map) */
-	if (Instance().FindBLEProvider(provider->GetBtAddress())) {
-		if (Instance().mRecovery.mAddressMatched) {
-			bt_addr_le_t address = provider->GetBtAddress();
-
-			if (memcmp(&Instance().mRecovery.mScannedDevice.mAddr, &address, sizeof(address)) == 0) {
-				DeviceLayer::PlatformMgr().ScheduleWork(
-					[](intptr_t context) {
-						Instance().Reconnect(
-							reinterpret_cast<BLEBridgedDeviceProvider *>(context));
-					},
-					reinterpret_cast<intptr_t>(provider));
-				return;
+	BLEBridgedDeviceProvider *provider;
+
+	sys_snode_t *node = sys_slist_peek_head(&Instance().mRecovery.mListToRecover);
+	Recovery::ListItem *item;
+	bool providerFound = false;
+	bt_addr_le_t providerAddress;
+
+	/* Filter out devices to recover from a scan result and put them to the queue to reconnect. */
+	do {
+		item = CONTAINER_OF(node, __typeof__(*item), mNode);
```

Comment:
Ditto. And also, if we cannot use Zephyr's API maybe we can create a short helper method to not duplicate a code?